sklearn/utils/validation.py

```diff
@@ -515,68 +515,150 @@ def _ensure_sparse_format(
                 stacklevel=2,
             )
         else:
             _assert_all_finite(
-                spmatrix.data,
+                sparse_container.data,
                 allow_nan=force_all_finite == "allow-nan",
                 estimator_name=estimator_name,
                 input_name=input_name,
             )
 
-    return spmatrix
+    # With SciPy sparse arrays, conversion from DIA format to COO, CSR, or BSR triggers
+    # the use of `np.int64` indices even if the data is such that it could be more
+    # efficiently represented with `np.int32` indices.
+    # https://github.com/scipy/scipy/issues/19245
+    # Since not all scikit-learn algorithms support large indices, the following code
+    # downcasts to `np.int32` indices when it's safe to do so.
+    if (
+        sparse_container_type_name == "dia_array"
+        and changed_format
+        and accept_sparse[0] in ("csr", "coo")
+    ):
+        if accept_sparse[0] == "csr":
+            index_dtype = _smallest_admissible_index_dtype(
+                arrays=(sparse_container.indptr, sparse_container.indices),
+                maxval=max(sparse_container.nnz, sparse_container.shape[1]),
+                check_contents=True,
+            )
+            if index_dtype != sparse_container.indices.dtype:
+                sparse_container.indices = sparse_container.indices.astype(index_dtype)
+                sparse_container.indptr = sparse_container.indptr.astype(index_dtype)
+        else:  # accept_sparse[0] == "coo"
+            index_dtype = _smallest_admissible_index_dtype(
+                maxval=max(sparse_container.shape)
+            )
+            if index_dtype != sparse_container.row.dtype:
+                sparse_container.row = sparse_container.row.astype(index_dtype)
+                sparse_container.col = sparse_container.col.astype(index_dtype)
+
+    return sparse_container
+
+
+def _smallest_admissible_index_dtype(arrays=(), maxval=None, check_contents=False):
+    """Based on input (integer) arrays `a`, determine a suitable index data
+    type that can hold the data in the arrays.
+
+    `max_values` prevails on arrays dtype and there contents.
+
+    Parameters
+    ----------
+    arrays : ndarray or tuple of ndarrays, default=()
+        Input arrays whose types/contents to check.
+
+    maxval : float, default=None
+        Maximum value needed.
+
+    check_contents : bool, default=False
+        Whether to check the values in the arrays and not just their types.
+        By default, check only the types.
+
+    Returns
+    -------
+    dtype : {np.int32, np.int64}
+        Suitable index data type (int32 or int64).
+    """
+
+    int32min = np.int32(np.iinfo(np.int32).min)
+    int32max = np.int32(np.iinfo(np.int32).max)
+
+    dtype = np.intc
+    if maxval is not None:
+        maxval = np.int64(maxval)
+        if maxval > int32max:
+            dtype = np.int64
```

Comment:
Since the rest of the code can only set dtype to `np.int64` based on the contents we can actually directly return here:

```suggestion
            return np.int64
```